coil-base/src/main/java/coil/interceptor/EngineInterceptor.kt

```diff
@@ -0,0 +1,333 @@
+package coil.interceptor
+
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.Drawable
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import coil.ComponentRegistry
+import coil.EventListener
+import coil.annotation.ExperimentalCoilApi
+import coil.bitmappool.BitmapPool
+import coil.decode.DataSource
+import coil.decode.DecodeUtils
+import coil.decode.DrawableDecoderService
+import coil.decode.EmptyDecoder
+import coil.decode.Options
+import coil.fetch.DrawableResult
+import coil.fetch.Fetcher
+import coil.fetch.SourceResult
+import coil.memory.MemoryCache
+import coil.memory.RealMemoryCache
+import coil.memory.RequestService
+import coil.memory.StrongMemoryCache
+import coil.memory.WeakMemoryCache
+import coil.request.ImageRequest
+import coil.request.Metadata
+import coil.request.RequestResult
+import coil.request.SuccessResult
+import coil.size.OriginalSize
+import coil.size.PixelSize
+import coil.size.Size
+import coil.transform.Transformation
+import coil.util.Logger
+import coil.util.SystemCallbacks
+import coil.util.Utils.REQUEST_TYPE_ENQUEUE
+import coil.util.allowInexactSize
+import coil.util.closeQuietly
+import coil.util.fetcher
+import coil.util.foldIndices
+import coil.util.invoke
+import coil.util.log
+import coil.util.mapData
+import coil.util.requireDecoder
+import coil.util.requireFetcher
+import coil.util.safeConfig
+import coil.util.set
+import coil.util.takeIf
+import coil.util.toDrawable
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.ensureActive
+import kotlin.coroutines.coroutineContext
+import kotlin.math.abs
+
+/** The last interceptor in the chain which executes the [ImageRequest]. */
+@OptIn(ExperimentalCoilApi::class)
+internal class EngineInterceptor(
```

Comment:
This is a combination of existing code from `RealImageLoader` and `MemoryCacheService`. It takes in an `ImageRequest` and returns a `RequestResult` without handling lifecycles or setting the result on the target (that's still handled in `RealImageLoader`.

Reply:
Not sure I understand the naming convention, but makes sense 👍

Reply:
Ah yeah I was going for a [software engine](https://en.wikipedia.org/wiki/Software_engine), though definitely could be improved (open to suggestions!). Fortunately, it's not public API so it can be changed in the future.

Reply:
Ahhh, nice, I like that.